Introduce toVNode() to reconstruct root element as vnode

staltz opened the pull request:
This commit addresses issue #167. 

Previously, in snabbdom v0.5.0, patch(element, vnode) would always create a new element for the root. This resulted in problems with custom elements (web components), and was fixed in commit c091c59.

However, that commit resulted in bug #167. This meant that snabbdom would have bugs with server-side rendering, where the root element would be non-empty (it has many children, rendered on the server-side as HTML), and the client-side rendering should reuse those existing children (or clear all the children and recreate them again in patch()).

This commit introduces the function toVNode(elm) that deep-converts an element (and its tree) to a VNode (and its tree), that is separately imported and used before calling patch(). toVNode(elm) will look at the element's attributes and gather those as data for the vnode.

Overall, this commit is important for fixing #167 and enabling client/server-side rendering in an efficient manner (destruction/recreation client-side is probably too expensive).

**Would appreciate reviews as soon as possible since this is a critical bug that is blocking development of a new version in Cycle.js**

nunocastromartins commented on src/snabbdom.ts:
```diff
@@ -57,10 +57,27 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
+      const node = elmChildren[i];
+      if (node.nodeType === 1) {
+        children.push(toVNode(elmChildren[i] as Element));
+      } else if (node.nodeType === 2) {
```
Did you test this? `Text` nodes should have a `nodeType` of `3`.
https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Node_type_constants

Also, I'd consider creating a `nodeType` function in `htmldomapi.ts` instead, to be consistent with how the `tagName` is accessed. Maybe also declare the `1` and `3` as constants there?

staltz replied:
👍

staltz commented:
Fixed now with @nunocastromartins 's suggestions.

AlexGalays commented on src/modules/attributes.ts:
```diff
@@ -26,28 +26,31 @@ function updateAttrs(oldVnode: VNode, vnode: VNode): void {
+    const lowkey = key.toLowerCase();
+    lowAttrs[lowkey] = cur;
+    old = oldAttrs[lowkey];
     if (old !== cur) {
```
For subsequent patchings (a VNode against a VNode), won't that compare the lowerCased current attribute with the non lowerCased previous attribute and thus do a wasteful setAttribute() ?

nunocastromartins commented on src/snabbdom.ts:
```diff
@@ -57,10 +57,29 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
+      } else if (api.isText(node)) {
+        children.push(
+          (vnode as any)(undefined, undefined, undefined, node.textContent)
+        );
```
Now that Snabbdom supports comment nodes, maybe you should also check for them (`nodeType` is `8`)?

nunocastromartins commented on src/snabbdom.ts:
```diff
@@ -57,10 +57,29 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
   }
 
-  function emptyNodeAt(elm: Element) {
+  function toVNode(elm: Element): VNode {
```
Why not make this a bit more generic and accept `Node` instead of `Element`? This would make it possible to patch a virtual node against a `Text` node (and since a recent PR, against `Comment` nodes too).

nunocastromartins commented on src/snabbdom.ts:
```diff
@@ -57,10 +57,29 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
+        children.push(toVNode(node));
+      } else if (api.isText(node)) {
+        children.push(
+          (vnode as any)(undefined, undefined, undefined, node.textContent)
```
I'd also put a `getTextContent` function in the API, in conformity with `setTextContent`.

caridy commented on src/snabbdom.ts:
```diff
@@ -57,10 +57,29 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
+    for (let i = 0, elmAttrs = elm.attributes, n = elmAttrs.length; i < n; i++) {
+      attrs[elmAttrs[i].nodeName] = elmAttrs[i].nodeValue;
+    }
+    delete attrs.id;
```
`delete` might have undesirable "perf" consequences (lead to deoptimization) in dictionary-like objects like `attrs`. Instead of adding the keys and then delete them immediately after, just opt-out of the addition of the key in the first place (maybe an `if` statement in line 67). Btw, I'm not sure this is a critical path for the engine where the `delete` will have an impact.

caridy commented on src/modules/attributes.ts:
```diff
@@ -26,28 +26,31 @@ function updateAttrs(oldVnode: VNode, vnode: VNode): void {
   for (key in attrs) {
     cur = attrs[key];
-    old = oldAttrs[key];
+    const lowkey = key.toLowerCase();
```
why lowercasing the keys in `attrs` when attribute keys are suppose to be in lowercase in the first place? snabbdom is a low-level API, and whoever relies on `attrs` must ensure that attribute keys (per spec) are lowercased. Seems to me that we are slowing down the critical path to accomodate a very edge case in which someone uses snabbdom incorrectly.

caridy replied:
if this is because `toVNode()` and not really about folks using snabbdom incorrectly, then the lowercasing should happen during `toVNode()` routine instead instead of doing it during the critical path.

AlexGalays replied:
Agreed. We can always add the most common attributes to the typings in lowercase form too, to limit errors (and people usually know that custom data attributes are in lower case). Personally, I will never use server-side rendering but with this change I would still have to pay the price of lowercasing every single attribute of my app, on every render (twice).

staltz replied:
My problem was this test https://github.com/snabbdom/snabbdom/blob/master/test/attributes.js#L65 which uses snabbdom "incorrectly" as you mentioned.

caridy replied:
then, lets clarify the api, where `attrs` is suppose to be key value pairs where the key is suppose to be a valid html attribute name (per spec), and let's fix the test to provide the right form.

staltz commented:
- Changed `toVNode(elm: Element): VNode` to `toVNode(elm: Node): VNode` as @nunocastromartins requested
- Made `toVNode` check nodes of type Comment too, @nunocastromartins 
- Created `DOMAPI.getTextContent` to match `setTextContent`, @nunocastromartins 
- Replaced `delete` with `if` blacklist, as @caridy requested

staltz commented:
Also, I reverted changes to the Attributes Module. I update the attributes test to avoid using camel-cased attributes.

staltz commented on test/attributes.js:
```diff
@@ -62,15 +74,11 @@ describe('attributes', function() {
   });
   describe('Object.prototype property', function() {
     it('is not considered as a boolean attribute and shouldn\'t be omitted', function() {
```
For the purpose of just checking Object.prototype properties, it's enough to test for the `constructor` property. Attributes like `valueOf` are invalid since they contain an uppercase letter.
Requesting @fix-fix to review this since they wrote this test.

fix-fix replied:
@staltz LGTM, even testing for single property is enough. Though it might be better to check using `true`/`false` instead of `'truthy'`/`null` for this single case to make it more obvious.

TylorS commented:
From a more general stance of type-safety, I think it'd be even better to have this function be required for the user to make use of, which in turn will allow patch to have a monomorphic type signature. For example

```typescript
const element = document.querySelector('#root') as Element

const rootVNode: VNode = toVNode(element)

patch(rootVNode, h('div', {}, [ ... ])
```

If anyone agrees with me, this *is* a breaking change, I think the types can be greatly improved from there as well, but I'll avoid going into that right now.

staltz commented:
Nice idea @TylorS. I would discuss that in a separate issue only after this PR is merged, so we don't delay merging this too much, but I like it.

TylorS commented:
Fain enough, opening a separate issue.

AlexGalays commented on src/snabbdom.ts:
```diff
@@ -60,10 +60,38 @@ export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
-    return vnode(api.tagName(elm).toLowerCase() + id + c, {}, [], undefined, elm);
+  function toVNode(elm: Node): VNode {
+    let text: string;
+    if (api.isText(elm)) {
```
Nitpicking but I suspect `toVNode` will be called with an Element rather than a text or comment, most of the time. Wouldn't it make more sense to start with `if (api.isElement(elm))`?

staltz replied:
That's smart, I'll do that

paldepind commented:
This is definitely an improvement when using server side rendering 😄 

I think I prefer keeping `emptyNodeAt` and moving `toVNode` to a seperate file. If one does not use server side rendering then `toVNode` isn't very useful. Furthermore it relies on the attributes module. Both are reasons for having it in a seperate file in my opinion.

Besides the above I think the PR is wonderful and a very welcome feature addition.

staltz commented:
Good point, I can move it to a separate file. However, it needs the `api` object. How do you recommend that `toVNode` depend on that? 'Optional with default', like `init` has?

paldepind commented:
'Optional with default' sounds good to me 👍

staltz commented:
I made `toVNode` a separate file, @paldepind.
Updated the attributes test to use true/false, @fix-fix.

staltz commented:
Thanks :)
